File: android/app/src/main/kotlin/com/player/tv/data/remote/StalkerApiService.kt
package com.player.tv.data.remote

import com.player.tv.domain.model.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.*
import javax.inject.Inject
import javax.inject.Singleton

@Serializable
data class StalkerChannel(
    val id: Int = 0,
    val name: String = "",
    val cmd: String = "",
    val tv_genre_id: Int = 0,
    val number: String = "",
    val censored: Int = 0,
    val archive_duration: Int = 0,
    val archive: Int = 0,
    val epg: StalkerEpg? = null,
    val t_shift: Int = 0,
    val x_id: Int = 0,
    val t_time: String = "",
    val t_time_to: String = "",
    val logo: String = "",
    val curtain: String = ""
)

@Serializable
data class StalkerEpg(
    val e_id: String = "",
    val name: String = "",
    val t_time: String = "",
    val t_time_to: String = "",
    val genre: String = ""
)

@Serializable
data class StalkerGenre(
    val id: Int = 0,
    val name: String = "",
    val alias: String = ""
)

@Serializable
data class StalkerMovie(
    val id: Int = 0,
    val name: String = "",
    val cmd: String = "",
    val year: String = "",
    val time: String = "",
    val tv_genre_id: Int = 0,
    val director: String = "",
    val actors: String = "",
    val rating_imdb: String = "",
    val description: String = "",
    val logo: String = "",
    val genres: List<StalkerGenre> = emptyList()
)

@Serializable
data class StalkerOrderedListResponse(
    val js: StalkerJs? = null
)

@Serializable
data class StalkerJs(
    val data: List<StalkerChannel> = emptyList(),
    val total_items: Int = 0,
    val per_page: Int = 0,
    val start_page: Int = 0,
    val cur_page: Int = 0,
    val max_page_items: Int = 0
)

@Singleton
class StalkerApiService @Inject constructor() {

    private val json = Json { ignoreUnknownKeys = true }

    suspend fun handshake(
        serverUrl: String,
        macAddress: String,
        deviceId: String = "",
        stbType: String = "STB"
    ): String? = withContext(Dispatchers.IO) {
        try {
            val url = "$serverUrl/stalker_portal/server/load.php?action=handshake&type=stb&token=&JsHttpRequest=1&ver=7.4.7&device_id=$deviceId&device_id2=$macAddress&stb_type=$stbType"
            val connection = java.net.URL(url).openConnection()
            connection.setRequestProperty("Cookie", "mac=$macAddress")
            connection.setRequestProperty("X-User-Agent", "PortalForStb")
            connection.connect()

            // Extract token from response
            val response = connection.getInputStream().bufferedReader().readText()
            val token = extractToken(response)

            // Get MAC from cookie if not provided
            val cookies = connection.getHeaderField("Set-Cookie")
            token
        } catch (e: Exception) {
            null
        }
    }

    private fun extractToken(response: String): String? {
        return try {
            val json = json.parseToJsonElement(response)
            json.jsonObject["js"]?.jsonObject?.get("token")?.jsonPrimitive?.content
        } catch (e: Exception) {
            null
        }
    }

    suspend fun getLiveChannels(
        serverUrl: String,
        token: String,
        macAddress: String,
        categoryId: Int? = null,
        page: Int = 1,
        perPage: Int = 50
    ): List<StalkerChannel> = withContext(Dispatchers.IO) {
        try {
            var url = "$serverUrl/stalker_portal/server/load.php?type=itv&retession_hours=24&action=get_ordered_list&p=$page&c=$perPage"
            if (categoryId != null) {
                url += "&genre=$categoryId"
            }
            val response = makeAuthenticatedRequest(url, token, macAddress)
            val result = json.decodeFromString<StalkerOrderedListResponse>(response)
            result.js?.data ?: emptyList()
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun getLiveCategories(
        serverUrl: String,
        token: String,
        macAddress: String
    ): List<StalkerGenre> = withContext(Dispatchers.IO) {
        try {
            val url = "$serverUrl/stalker_portal/server/load.php?type=itv&action=get_genres"
            val response = makeAuthenticatedRequest(url, token, macAddress)
            val result = json.parseToJsonElement(response)
            val data = result.jsonObject["js"]?.jsonArray ?: return@withContext emptyList()
            data.map { element ->
                json.decodeFromJsonElement<StalkerGenre>(element)
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun getVodMovies(
        serverUrl: String,
        token: String,
        macAddress: String,
        categoryId: Int? = null,
        page: Int = 1,
        perPage: Int = 50
    ): List<StalkerMovie> = withContext(Dispatchers.IO) {
        try {
            var url = "$serverUrl/stalker_portal/server/load.php?type=vod&retession_hours=24&action=get_ordered_list&p=$page&c=$perPage"
            if (categoryId != null) {
                url += "&genre=$categoryId"
            }
            val response = makeAuthenticatedRequest(url, token, macAddress)
            val result = json.parseToJsonElement(response)
            val data = result.jsonObject["js"]?.jsonObject?.get("data")?.jsonArray ?: return@withContext emptyList()
            data.map { element ->
                json.decodeFromJsonElement<StalkerMovie>(element)
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun getVodCategories(
        serverUrl: String,
        token: String,
        macAddress: String
    ): List<StalkerGenre> = withContext(Dispatchers.IO) {
        try {
            val url = "$serverUrl/stalker_portal/server/load.php?type=vod&action=get_genres"
            val response = makeAuthenticatedRequest(url, token, macAddress)
            val result = json.parseToJsonElement(response)
            val data = result.jsonObject["js"]?.jsonArray ?: return@withContext emptyList()
            data.map { element ->
                json.decodeFromJsonElement<StalkerGenre>(element)
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun createLiveLink(
        serverUrl: String,
        token: String,
        macAddress: String,
        channelId: Int
    ): String? = withContext(Dispatchers.IO) {
        try {
            val url = "$serverUrl/stalker_portal/server/load.php?type=itv&action=create_link&ch_id=$channelId"
            val response = makeAuthenticatedRequest(url, token, macAddress)
            val result = json.parseToJsonElement(response)
            result.jsonObject["js"]?.jsonObject?.get("cmd")?.jsonPrimitive?.content
        } catch (e: Exception) {
            null
        }
    }

    suspend fun createVodLink(
        serverUrl: String,
        token: String,
        macAddress: String,
        movieId: Int
    ): String? = withContext(Dispatchers.IO) {
        try {
            val url = "$serverUrl/stalker_portal/server/load.php?type=vod&action=create_link&movie_id=$movieId"
            val response = makeAuthenticatedRequest(url, token, macAddress)
            val result = json.parseToJsonElement(response)
            result.jsonObject["js"]?.jsonObject?.get("cmd")?.jsonPrimitive?.content
        } catch (e: Exception) {
            null
        }
    }

    suspend fun getChannelEpg(
        serverUrl: String,
        token: String,
        macAddress: String,
        channelId: Int
    ): List<StalkerEpg> = withContext(Dispatchers.IO) {
        try {
            val url = "$serverUrl/stalker_portal/server/load.php?type=itv&action=get_short_epg&ch_id=$channelId&P=1&L=24"
            val response = makeAuthenticatedRequest(url, token, macAddress)
            val result = json.parseToJsonElement(response)
            val epgs = result.jsonObject["js"]?.jsonObject?.get("epg_listings")?.jsonArray ?: return@withContext emptyList()
            epgs.map { element ->
                json.decodeFromJsonElement<StalkerEpg>(element)
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    private fun makeAuthenticatedRequest(url: String, token: String, macAddress: String): String {
        val connection = java.net.URL(url).openConnection()
        connection.setRequestProperty("Cookie", "mac=$macAddress;stb_token=$token")
        connection.setRequestProperty("Authorization", "Bearer $token")
        connection.setRequestProperty("X-User-Agent", "PortalForStb")
        connection.connect()
        return connection.getInputStream().bufferedReader().readText()
    }

    fun convertToChannel(stalkerChannel: StalkerChannel): Channel {
        return Channel(
            id = "stalker_${stalkerChannel.id}",
            name = stalkerChannel.name,
            url = "", // Will be filled when creating link
            logo = stalkerChannel.logo.ifEmpty { null },
            groupTitle = null,
            tvgId = null,
            tvgName = stalkerChannel.name,
            isRadio = false,
            streamFormat = StreamFormat.HLS,
            streamType = StreamType.LIVE
        )
    }

    fun convertToChannel(stalkerMovie: StalkerMovie): Channel {
        return Channel(
            id = "stalker_vod_${stalkerMovie.id}",
            name = stalkerMovie.name,
            url = "", // Will be filled when creating link
            logo = stalkerMovie.logo.ifEmpty { null },
            groupTitle = stalkerMovie.genres.firstOrNull()?.name,
            tvgId = null,
            tvgName = stalkerMovie.name,
            isRadio = false,
            streamFormat = StreamFormat.HLS,
            streamType = StreamType.VOD
        )
    }
}

File: ios/Core/Services/StalkerApiService.swift
import Foundation

class StalkerApiService {
    
    private let decoder = JSONDecoder()
    
    struct StalkerChannel: Codable {
        let id: Int
        let name: String
        let cmd: String
        let tvGenreId: Int
        let number: String
        let logo: String?
        
        enum CodingKeys: String, CodingKey {
            case id, name, cmd, number, logo
            case tvGenreId = "tv_genre_id"
        }
    }
    
    struct StalkerGenre: Codable {
        let id: Int
        let name: String
    }
    
    struct StalkerMovie: Codable {
        let id: Int
        let name: String
        let cmd: String
        let year: String?
        let time: String?
        let description: String?
        let logo: String?
    }
    
    struct StalkerOrderedListResponse: Codable {
        let js: StalkerJs?
    }
    
    struct StalkerJs: Codable {
        let data: [StalkerChannel]?
        let totalItems: Int?
        
        enum CodingKeys: String, CodingKey {
            case data
            case totalItems = "total_items"
        }
    }
    
    func handshake(serverUrl: String, macAddress: String, deviceId: String = "", stbType: String = "STB") async throws -> String? {
        let urlString = "\(serverUrl)/stalker_portal/server/load.php?action=handshake&type=stb&token=&JsHttpRequest=1&ver=7.4.7&device_id=\(deviceId)&device_id2=\(macAddress)&stb_type=\(stbType)"
        guard let url = URL(string: urlString) else { return nil }
        
        var request = URLRequest(url: url)
        request.setValue("mac=\(macAddress)", forHTTPHeaderField: "Cookie")
        request.setValue("PortalForStb", forHTTPHeaderField: "X-User-Agent")
        
        let (data, _) = try await URLSession.shared.data(for: request)
        let response = try decoder.decode([String: AnyCodable].self, from: data)
        
        if let js = response["js"]?.value as? [String: Any],
           let token = js["token"] as? String {
            return token
        }
        
        return nil
    }
    
    func getLiveChannels(serverUrl: String, token: String, macAddress: String, categoryId: Int? = nil, page: Int = 1, perPage: Int = 50) async throws -> [StalkerChannel] {
        var urlString = "\(serverUrl)/stalker_portal/server/load.php?type=itv&retession_hours=24&action=get_ordered_list&p=\(page)&c=\(perPage)"
        if let categoryId = categoryId {
            urlString += "&genre=\(categoryId)"
        }
        guard let url = URL(string: urlString) else { return [] }
        
        var request = URLRequest(url: url)
        request.setValue("mac=\(macAddress);stb_token=\(token)", forHTTPHeaderField: "Cookie")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("PortalForStb", forHTTPHeaderField: "X-User-Agent")
        
        let (data, _) = try await URLSession.shared.data(for: request)
        let response = try decoder.decode(StalkerOrderedListResponse.self, from: data)
        
        return response.js?.data ?? []
    }
    
    func getLiveCategories(serverUrl: String, token: String, macAddress: String) async throws -> [StalkerGenre] {
        let urlString = "\(serverUrl)/stalker_portal/server/load.php?type=itv&action=get_genres"
        guard let url = URL(string: urlString) else { return [] }
        
        var request = URLRequest(url: url)
        request.setValue("mac=\(macAddress);stb_token=\(token)", forHTTPHeaderField: "Cookie")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("PortalForStb", forHTTPHeaderField: "X-User-Agent")
        
        let (data, _) = try await URLSession.shared.data(for: request)
        let response = try decoder.decode([String: AnyCodable].self, from: data)
        
        if let js = response["js"]?.value as? [[String: Any]] {
            let jsonData = try JSONSerialization.data(withJSONObject: js)
            return try decoder.decode([StalkerGenre].self, from: jsonData)
        }
        
        return []
    }
    
    func getVodMovies(serverUrl: String, token: String, macAddress: String, categoryId: Int? = nil, page: Int = 1, perPage: Int = 50) async throws -> [StalkerMovie] {
        var urlString = "\(serverUrl)/stalker_portal/server/load.php?type=vod&retession_hours=24&action=get_ordered_list&p=\(page)&c=\(perPage)"
        if let categoryId = categoryId {
            urlString += "&genre=\(categoryId)"
        }
        guard let url = URL(string: urlString) else { return [] }
        
        var request = URLRequest(url: url)
        request.setValue("mac=\(macAddress);stb_token=\(token)", forHTTPHeaderField: "Cookie")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("PortalForStb", forHTTPHeaderField: "X-User-Agent")
        
        let (data, _) = try await URLSession.shared.data(for: request)
        let response = try decoder.decode([String: AnyCodable].self, from: data)
        
        if let js = response["js"]?.value as? [String: Any],
           let dataArray = js["data"] as? [[String: Any]] {
            let jsonData = try JSONSerialization.data(withJSONObject: dataArray)
            return try decoder.decode([StalkerMovie].self, from: jsonData)
        }
        
        return []
    }
    
    func createLiveLink(serverUrl: String, token: String, macAddress: String, channelId: Int) async throws -> String? {
        let urlString = "\(serverUrl)/stalker_portal/server/load.php?type=itv&action=create_link&ch_id=\(channelId)"
        guard let url = URL(string: urlString) else { return nil }
        
        var request = URLRequest(url: url)
        request.setValue("mac=\(macAddress);stb_token=\(token)", forHTTPHeaderField: "Cookie")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("PortalForStb", forHTTPHeaderField: "X-User-Agent")
        
        let (data, _) = try await URLSession.shared.data(for: request)
        let response = try decoder.decode([String: AnyCodable].self, from: data)
        
        if let js = response["js"]?.value as? [String: Any],
           let cmd = js["cmd"] as? String {
            return cmd
        }
        
        return nil
    }
    
    func convertToChannel(_ stalkerChannel: StalkerChannel) -> Channel {
        return Channel(
            id: "stalker_\(stalkerChannel.id)",
            name: stalkerChannel.name,
            url: "",
            logo: stalkerChannel.logo,
            groupTitle: nil,
            tvgId: nil,
            tvgName: stalkerChannel.name,
            isRadio: false,
            streamFormat: .hls,
            streamType: .live
        )
    }
    
    func convertToChannel(_ stalkerMovie: StalkerMovie) -> Channel {
        return Channel(
            id: "stalker_vod_\(stalkerMovie.id)",
            name: stalkerMovie.name,
            url: "",
            logo: stalkerMovie.logo,
            groupTitle: nil,
            tvgId: nil,
            tvgName: stalkerMovie.name,
            isRadio: false,
            streamFormat: .hls,
            streamType: .vod
        )
    }
}

// Helper for decoding dynamic JSON
struct AnyCodable: Codable {
    let value: Any
    
    init(_ value: Any) {
        self.value = value
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let intVal = try? container.decode(Int.self) {
            value = intVal
        } else if let doubleVal = try? container.decode(Double.self) {
            value = doubleVal
        } else if let boolVal = try? container.decode(Bool.self) {
            value = boolVal
        } else if let stringVal = try? container.decode(String.self) {
            value = stringVal
        } else if let arrayVal = try? container.decode([AnyCodable].self) {
            value = arrayVal.map { $0.value }
        } else if let dictVal = try? container.decode([String: AnyCodable].self) {
            value = dictVal.mapValues { $0.value }
        } else {
            value = NSNull()
        }
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        if let intVal = value as? Int {
            try container.encode(intVal)
        } else if let doubleVal = value as? Double {
            try container.encode(doubleVal)
        } else if let boolVal = value as? Bool {
            try container.encode(boolVal)
        } else if let stringVal = value as? String {
            try container.encode(stringVal)
        } else {
            try container.encodeNil()
        }
    }
}
